### PR TITLE
[v0.33] Feat / OKEx connector

### PR DESCRIPTION
**Pull request by vic-en**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**After creating this PR, please make sure:**

- [ ] You link the related GitHub issue or ticket

**A description of the changes proposed in the pull request**:
This PR supersedes  #2262

### CODE REVIEW

File: hummingbot/connector/exchange/okex/okex_auth.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            "op": "login",
+            "args": [
+                self.api_key,
+                self.passphrase,
+                timestamp,
+                self.get_signature(timestamp, "GET", "/users/self/verify", {})
+            ]
+        }
+
+# auth = OKExAuth(okex_api_key, okex_secret_key, okex_passphrase)
```
Comment (Nullably) [inline hummingbot/connector/exchange/okex/okex_auth.py:71]: Should we remove these comments?

File: hummingbot/connector/exchange/okex/okex_api_order_book_data_source.py [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+            while True:
+                try:
+                    msg: str = await asyncio.wait_for(ws.recv(), timeout=self.MESSAGE_TIMEOUT)
+                    yield msg
+                except asyncio.TimeoutError:
+                    # try:
+                    pong_waiter = await ws.ping()
+                    await asyncio.wait_for(pong_waiter, timeout=self.PING_TIMEOUT)
+                    # removed this try as it would fail anyway, #TODO check this
+                    # except asyncio.TimeoutError:
```
Comment (Nullably) [inline hummingbot/connector/exchange/okex/okex_api_order_book_data_source.py:236]: Yep, you can remove this try except.

File: hummingbot/connector/exchange/okex/okex_exchange.pyx [hunk 3]
```diff
@@ -0,0 +1,1008 @@
...
+                    f"Error fetching status update for the order {tracked_order.client_order_id}: "
+                    f"{order_update}.",
+                    app_warning_msg=f"Could not fetch updates for the order {tracked_order.client_order_id}. "
+                                    f"The order has either been filled or canceled."
+                )
+                continue
+
+            # Calculate the newly executed amount for this update.
+            tracked_order.last_state = order_update["state"]
+            new_confirmed_amount = Decimal(order_update["filled_size"])  # TODO filled_notional or filled_size?
```
Comment (Nullably) [inline hummingbot/connector/exchange/okex/okex_exchange.pyx:435]: Is this TODO resolved?

File: hummingbot/connector/exchange/okex/okex_exchange.pyx [hunk 4]
```diff
@@ -0,0 +1,1008 @@
...
+            execute_amount_diff = new_confirmed_amount - tracked_order.executed_amount_base
+
+            if execute_amount_diff > s_decimal_0:
+                tracked_order.executed_amount_base = new_confirmed_amount
+                tracked_order.executed_amount_quote = Decimal(order_update["filled_notional"])  # TODO filled_notional or filled_size?
```
Comment (Nullably) [inline hummingbot/connector/exchange/okex/okex_exchange.pyx:440]: Is this TODO resolved?

File: hummingbot/connector/exchange/okex/okex_exchange.pyx [hunk 5]
```diff
@@ -0,0 +1,1008 @@
...
+        if price == s_decimal_0:
+            notional_size = current_price * quantized_amount
+        else:
+            notional_size = price * quantized_amount
+        # Add 1% as a safety factor in case the prices changed while making the order.
+
+        if notional_size < trading_rule.min_notional_size * Decimal("1.01"):
+            return s_decimal_0
+
+        return quantized_amount
```
Comment (Nullably) [inline hummingbot/connector/exchange/okex/okex_exchange.pyx:1008]: There should be def functions for all public cdef functions, e.g. buy(), sell(). See last section of `binance_exchange.pyx` for an example.

File: hummingbot/connector/exchange/okex/okex_order_book_tracker.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+                raise
+            except Exception:
+                self.logger().network(
+                    f"Unexpected error tracking order book for {trading_pair}.",
+                    exc_info=True,
+                    app_warning_msg="Unexpected error tracking order book. Retrying after 5 seconds."
+                )
+                await asyncio.sleep(5.0)
+
+    # def start(self):
```
Comment (Nullably) [inline hummingbot/connector/exchange/okex/okex_order_book_tracker.py:127]: Remove this?

**Review by Nullably — CHANGES_REQUESTED** [review]

File: hummingbot/connector/exchange/okex/okex_auth.py (hunk 1 above)
Comment (vic-en) [reply hummingbot/connector/exchange/okex/okex_auth.py:71]: Yes, I think it was initially used by the initial developer to test the OKExAuth class.

File: hummingbot/connector/exchange/okex/okex_api_order_book_data_source.py (hunk 2 above)
Comment (vic-en) [reply hummingbot/connector/exchange/okex/okex_api_order_book_data_source.py:236]: I think that is referring to the try except that was commented out.
Both actual try except on line 226 and 228 should remain according to instruction from the documentation page.

File: hummingbot/connector/exchange/okex/okex_exchange.pyx (hunk 4 above)
Comment (vic-en) [reply hummingbot/connector/exchange/okex/okex_exchange.pyx:440]: Yes.

File: hummingbot/connector/exchange/okex/okex_exchange.pyx (hunk 5 above)
Comment (vic-en) [reply hummingbot/connector/exchange/okex/okex_exchange.pyx:1008]: I did already, I forgot to push the commit.

File: hummingbot/connector/exchange/okex/okex_order_book_tracker.py (hunk 6 above)
Comment (vic-en) [reply hummingbot/connector/exchange/okex/okex_order_book_tracker.py]: Yes.

File: hummingbot/connector/exchange/okex/okex_exchange.pyx (hunk 3 above)
Comment (vic-en) [reply hummingbot/connector/exchange/okex/okex_exchange.pyx]: Yes.

File: hummingbot/connector/exchange/okex/okex_api_order_book_data_source.py (hunk 2 above)
Comment (Nullably) [reply hummingbot/connector/exchange/okex/okex_api_order_book_data_source.py]: Yeah, that's what i meant, to remove the commented try-except.

**Comment by rxlxrxsx** [comment]
On arbitrage strategy sometimes I get an ``Error running performance analysis`` when entering the history command. Also, ``Unexpected error running clock tick`` happens randomly when running the strategy.

[logs_okex-arb.log](https://github.com/CoinAlpha/hummingbot/files/5363432/logs_okex-arb.log)

Also, typo based on the logs. Filed instead of filled.

```
2020-10-12 05:48:11,804 - 15610 - hummingbot.connector.exchange.okex.okex_exchange - INFO - Filed 0.0667 out of 0.066700 
of order LIMIT-HUMMINGBOT1602481691010707
```

**Comment by rxlxrxsx** [comment]
Open orders are not getting canceled properly after entering the ``stop`` command thus leaving the orders stuck on the exchange. This is both happening on a single or multiple order level.

image
image

[logs_okex-test.log](https://github.com/CoinAlpha/hummingbot/files/5363624/logs_okex-test.log)

**Comment by vic-en** [comment]
@rxlxrxsx 
Kindly confirm fix now.
For arbitrage, run with any of the stable connectors like binance.

**Comment by RC-13** [comment]
`Unexpected error running clock tick` still persist when running arbitrage strategy
image

- No error encountered running History command
- Confirm typo for filled is now fixed

**Review by dennisocana — DISMISSED** [review]
QA confirmed all issues found during testing were fixed and can now be merged to `development` branch.

File: test/integration/test_okex_auth.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
+# import asyncio
+# import json
```
Comment (martinvol) [inline test/integration/test_okex_auth.py:2]:
```suggestion
```

File: test/integration/test_okex_auth.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+import unittest
+# from typing import List
+
+# import websockets
```
Comment (martinvol) [inline test/integration/test_okex_auth.py:6]:
```suggestion
```

File: hummingbot/connector/exchange/okex/constants.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+OKEX_PLACE_ORDER = "api/spot/v3/orders"
+OKEX_ORDER_DETAILS_URL = 'api/spot/v3/orders/{exchange_order_id}'
+OKEX_ORDER_CANCEL = 'api/spot/v3/cancel_orders/{exchange_order_id}'
+OKEX_BATCH_ORDER_CANCELL = 'api/spot/v3/cancel_batch_orders'
+OKEX_BALANCE_URL = "api/spot/v3/accounts"
+
+
+# WS
+OKCOIN_WS_URI = "wss://real.okex.com:8443/ws/v3"
```
Comment (martinvol) [inline hummingbot/connector/exchange/okex/constants.py:28]:
```suggestion
# WS
OKEX_WS_URI = "wss://real.okex.com:8443/ws/v3"
```

File: hummingbot/connector/exchange/okex/okex_api_order_book_data_source.py [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+from hummingbot.core.data_type.order_book import OrderBook
+from hummingbot.core.data_type.order_book_tracker_data_source import OrderBookTrackerDataSource
+from hummingbot.core.utils import async_ttl_cache
+from hummingbot.logger import HummingbotLogger
+from hummingbot.connector.exchange.okex.okex_order_book import OkexOrderBook
+from hummingbot.connector.exchange.okex.constants import (
+    OKEX_SYMBOLS_URL,
+    OKEX_PRICE_URL,
+    OKEX_DEPTH_URL,
+    OKCOIN_WS_URI,
```
Comment (martinvol) [inline hummingbot/connector/exchange/okex/okex_api_order_book_data_source.py:33]:
```suggestion
    OKEX_WS_URI,
```

File: hummingbot/connector/exchange/okex/okex_api_order_book_data_source.py [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+    def iso_to_timestamp(cls, date: str):
+        return dataparse(date).timestamp()
+
+    async def listen_for_trades(self, ev_loop: Optional[asyncio.BaseEventLoop], output: asyncio.Queue):
+        """Subscribes to the trade channel of the exchange. Adds incoming messages(of filled orders) to the output queue, to be processed by"""
+
+        while True:
+            try:
+                trading_pairs: List[str] = self._trading_pairs
+                async with websockets.connect(OKCOIN_WS_URI) as ws:
```
Comment (martinvol) [inline hummingbot/connector/exchange/okex/okex_api_order_book_data_source.py:186]:
```suggestion
                async with websockets.connect(OKEX_WS_URI) as ws:
```

File: hummingbot/connector/exchange/okex/okex_api_order_book_data_source.py [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+            return
+        finally:
+            await ws.close()
+
+    async def listen_for_order_book_diffs(self, ev_loop: Optional[asyncio.BaseEventLoop], output: asyncio.Queue):
+        """Fetches or Subscribes to the order book snapshots for each trading pair. Additionally, parses the incoming message into a OrderBookMessage and appends it into the output Queue."""
+        while True:
+            try:
+                trading_pairs: List[str] = await self.get_trading_pairs()
+                async with websockets.connect(OKCOIN_WS_URI) as ws:
```
Comment (martinvol) [inline hummingbot/connector/exchange/okex/okex_api_order_book_data_source.py:247]:
```suggestion
                async with websockets.connect(OKEX_WS_URI) as ws:
```

File: hummingbot/connector/exchange/okex/okex_api_user_stream_data_source.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    AsyncIterable,
+    List
+)
+
+from hummingbot.core.data_type.user_stream_tracker_data_source import UserStreamTrackerDataSource
+from hummingbot.logger import HummingbotLogger
+from hummingbot.connector.exchange.okex.okex_auth import OKExAuth
+
+from hummingbot.connector.exchange.okex.okex_utils import inflate
+from hummingbot.connector.exchange.okex.constants import OKCOIN_WS_URI
```
Comment (martinvol) [inline hummingbot/connector/exchange/okex/okex_api_user_stream_data_source.py:19]:
```suggestion
from hummingbot.connector.exchange.okex.constants import OKEX_WS_URI
```

File: hummingbot/connector/exchange/okex/okex_api_user_stream_data_source.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+        await self._websocket_connection.send(request)
+        resp = await self._websocket_connection.recv()
+        msg = json.loads(inflate(resp))
+        if msg["event"] != "subscribe":
+            self.logger().error(f"Error occurred subscribing to topic. {topic}. {msg}")
+        self.logger().info(f"Successfully subscribed to {topic}")
+
+    async def get_ws_connection(self):
+
+        stream_url: str = f"{OKCOIN_WS_URI}"
```
Comment (martinvol) [inline hummingbot/connector/exchange/okex/okex_api_user_stream_data_source.py:79]:
```suggestion
        stream_url: str = f"{OKEX_WS_URI}"
```

**Review by Nullably — APPROVED** [review]
LGTM

**Pull request merged**